### PR TITLE
Create interface to manage CalculatorConstants

### PR DESCRIPTION
**Pull request by hughes-dustin**
## Description
https://app.zenhub.com/workspaces/vft-59c95ae5fda7577a9b3184f8/issues/department-of-veterans-affairs/va.gov-team/6163

Allows GIDS users to manually update Calculator Constants that are used in the vets-website comparison tool. 

This PR is dependent upon [PR 606](https://github.com/department-of-veterans-affairs/gibct-data-service/pull/606) being merged into master.

## Testing done
Dev tested.

## Screenshots
image
image

On save:
image

## Acceptance criteria
- [x] There is a "Calculator Constants" option in the header menu/nav bar at the top of the page
- [x] When clicking "Calculator Constants", the user is redirected to a new page that contains a form for managing the calculator constants.
     - The page title is: "Calculator Constants"
     - There is a description of how the data is used.
- [x] The form displays a read only view of the following data:
     - Constant Name
     - Constant Description (See supporting artifact 2)
- [x] The form allows the user to edit the following data:
     - Constant Value
- [x] The calculator constant meets the the following criteria:
     - Is required
     - Is a number (need to differentiate fields between integers v. floats)
- [x] The user can cancel editing the form and it returns them to the dashboard.
- [x] The user can save the form edits and they are persisted.
- [x] If there is an error, no changes are made to the data.

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: app/assets/javascripts/calculator_constants.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
+const CONSTANT_DESCRIPTIONS = {
```
Comment (zurbergram) [inline app/assets/javascripts/calculator_constants.js]: why not put these in the database table?

File: app/controllers/calculator_constants_controller.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    end
+    flash.notice = 'Calculator Constants have been updated.'
+    redirect_to action: :index
+  end
+end
+
+private
+
+def update_upload_timestamp(upload, updated_timestamp)
+  upload.update(updated_at: updated_timestamp, created_at: updated_timestamp) if updated_timestamp > upload.updated_at
```
Comment (zurbergram) [inline app/controllers/calculator_constants_controller.rb]: This shouldn't update the `created_at` column, additionally `completed_at` and `user_id` should be updated.

Can you update the `comment` column as well to indicate what fields were updated or a generic message to indicate that CalculatorConstants were updated through this process?

File: app/controllers/calculator_constants_controller.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+class CalculatorConstantsController < ApplicationController
+  def index
+    @calculator_constants = CalculatorConstant.all
+  end
+
+  def update
+    upload = Upload.where(csv_type: 'CalculatorConstant').first
```
Comment (zurbergram) [inline app/controllers/calculator_constants_controller.rb]: See https://github.com/department-of-veterans-affairs/gibct-data-service/blob/master/app/models/upload.rb#L72 for how to pull most recent Upload row

```suggestion
    upload = Upload.select('DISTINCT ON("csv_type") *').where({csv_type: 'CalculatorConstant', ok: true}).order(updated_at: :desc)
```

File: app/controllers/calculator_constants_controller.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+class CalculatorConstantsController < ApplicationController
+  def index
+    @calculator_constants = CalculatorConstant.all
```
Comment (zurbergram) [inline app/controllers/calculator_constants_controller.rb]: Is any ordering needed?

File: spec/controllers/calculator_constants_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'support/controller_macros'
+require 'support/devise'
+require 'controllers/shared_examples/shared_examples_for_authentication'
+
+RSpec.describe CalculatorConstantsController, type: :controller do
```
Comment (zurbergram) [inline spec/controllers/calculator_constants_controller_spec.rb]: Need to add something like below to use the `shared_examples_for_authentication`
```
  it_behaves_like 'an authenticating controller', :partials, 'calculator_constants'
```

File: app/assets/javascripts/calculator_constants.js (hunk 1 above)
Comment (hughes-dustin) [reply app/assets/javascripts/calculator_constants.js]: Was waiting on feedback on where those descriptions were coming from. It's in the DB now.

File: app/controllers/calculator_constants_controller.rb (hunk 2 above)
Comment (hughes-dustin) [reply app/controllers/calculator_constants_controller.rb]: The `created_at` is what is being used in  the upload table on the dashboard. One of the requirements for this story was to update that displayed date when changes are made. 

I'll go ahead and add the changes to update the other two fields as well as create a generic message for the comment column.

File: app/controllers/calculator_constants_controller.rb (hunk 4 above)
Comment (hughes-dustin) [reply app/controllers/calculator_constants_controller.rb]: No ordering was requested.

File: spec/controllers/calculator_constants_controller_spec.rb (hunk 5 above)
Comment (hughes-dustin) [reply spec/controllers/calculator_constants_controller_spec.rb]: Added it in.

File: app/controllers/calculator_constants_controller.rb (hunk 3 above)
Comment (hughes-dustin) [reply app/controllers/calculator_constants_controller.rb]: Code's been removed!

File: app/controllers/calculator_constants_controller.rb (hunk 2 above)
Comment (hughes-dustin) [reply app/controllers/calculator_constants_controller.rb]: upload is no longer being updated by calculator constants management table.

**Review by shawkey-daniel — APPROVED** [review]
LGTM

File: app/views/layouts/_navbar.html.erb [hunk 6]
```diff
@@ -23,6 +23,7 @@
         <li class="<%= li_active_class(crosswalk_issues_partials_path) %>"><%= link_if_not_active 'Crosswalk Partial Matches', crosswalk_issues_partials_path %></li>
         <li class="<%= li_active_class(crosswalk_issues_orphans_path) %>"><%= link_if_not_active 'Ipeds Orphans', crosswalk_issues_orphans_path %></li>
         <li class="<%= li_active_class(archives_path) %>"><%= link_if_not_active 'Archives', archives_path %></li>
+        <li class="<%= li_active_class(calculator_constants_path) %>"><%= link_if_not_active 'Calculator Constants', calculator_constants_path  %></li>
```
Comment (zurbergram) [inline app/views/layouts/_navbar.html.erb]: Now that https://github.com/department-of-veterans-affairs/gibct-data-service/blob/master/config/initializers/common.rb#L18 is available should this be wrapped in a `if !production?` check?

Comment (hughes-dustin) [reply app/views/layouts/_navbar.html.erb]: See dev channel on slack.

Comment (hughes-dustin) [reply app/views/layouts/_navbar.html.erb]: No.

**Review by sonntag-adam — APPROVED** [review]

**Comment by omgitsbillryan** [comment]
Your link to PR 606 in the description is broken.

**Comment by shawkey-daniel** [comment]
Updated the link to PR 606

**Review by annaswims — APPROVED** [review]
LGTM

**Comment by annaswims** [comment]
please make sure the migration in https://github.com/department-of-veterans-affairs/gibct-data-service/pull/606 is run before this is deployed

File: app/controllers/calculator_constants_controller.rb [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  end
+
+  def update
+    updated_fields = []
+    params.each do |key, value|
+      update_calculator_constant(key, value, updated_fields)
+    end
```
Comment (omgitsbillryan) [inline app/controllers/calculator_constants_controller.rb]: `params` is going to contain things like `"controller"=>"calculator_constants", "action"=>"update"`. I'd recommend you make keys & values be scoped to change this API to accept a single input param var, i.e.
```
"calculator_constants" => [
  {"name": "foo", "float_value": 999.9},
  {"name": "bar", "float_value": 999.9}
]
```

**Review by omgitsbillryan — COMMENTED** [review]
I noticed that `app/controllers/v0/calculator_constants_controller.rb` already exists, but you've created a brand new `calculator_constants_controller.rb` w/o the `/v0/` api versioning. Is this intentional?

**Review by omgitsbillryan — CHANGES_REQUESTED** [review]
Gonna play bad cop for now - have some comments above ☝️

File: app/controllers/calculator_constants_controller.rb [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+      }
+    end
+    redirect_to action: :index
+  end
+
+  private
+
+  def update_calculator_constant(name, value, updated_fields)
+    if CalculatorConstant.exists?(name: name)
+      constant_field = CalculatorConstant.where(name: name).first
```
Comment (omgitsbillryan) [inline app/controllers/calculator_constants_controller.rb]: This is going to result in a lot of SQL calls. I think you can instead be more performant by a single query:

```ruby
CalculatorConstant.where(name: ["AVGDODBAH","IGMCKY",...])
```

File: sample_csvs/calculator_constant.csv [hunk 9]
```diff
@@ -1,29 +1,29 @@
...
+AVGVABAH,1681.0	AVGVABAH,1681,Monthly Housing Allowance
+AVGDODBAH,1650.0	AVGDODBAH,1650,Average Monthly Housing Allowance
+DEARATEFULLTIME,1224.0	DEARATEFULLTIME,1224,Educational Assistance Allowance for trainees under the Survivors' and Dependents' Educational Assistance Program (Chapter 35 of title 38, U.S.C.) Full Time
+DEARATETHREEQUARTERS,967.0	DEARATETHREEQUARTERS,967,Educational Assistance Allowance for trainees under the Survivors' and Dependents' Educational Assistance Program (Chapter 35 of title 38, U.S.C.) 3/4
+DEARATEONEHALF,710.0	DEARATEONEHALF,710,Educational Assistance Allowance for trainees under the Survivors' and Dependents' Educational Assistance Program (Chapter 35 of title 38, U.S.C.) 1/2
+DEARATEUPTOONEHALF,710.0	DEARATEUPTOONEHALF,710,Educational Assistance Allowance for trainees under the Survivors' and Dependents' Educational Assistance Program (Chapter 35 of title 38, U.S.C.) Up to a Quarter
+DEARATEUPTOONEQUARTER,306.0	DEARATEUPTOONEQUARTER,306,Educational Assistance Allowance for trainees under the Survivors' and Dependents' Educational Assistance Program (Chapter 35 of title 38, U.S.C.) 1 half
+FISCALYEAR,2019.0	FISCALYEAR,2019,Fiscal Year
```
[the remark this answers is not in the document]
Comment (hughes-dustin) [reply sample_csvs/calculator_constant.csv]: This looks like the aftermath of a bad merge from master. I've pushed in the fix.

File: app/controllers/calculator_constants_controller.rb (hunk 7 above)
Comment (hughes-dustin) [reply app/controllers/calculator_constants_controller.rb]: Made the change. Also learned something about how to make an array of hashes for a form submit. Neat trick.

File: app/controllers/calculator_constants_controller.rb (hunk 8 above)
Comment (hughes-dustin) [reply app/controllers/calculator_constants_controller.rb]: Good point. I made the change. Thans!

File: app/controllers/calculator_constants_controller.rb [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    end
+    redirect_to action: :index
+  end
+
+  private
+
+  def update_calculator_constant(submitted_constants, updated_fields)
+    constant_fields = CalculatorConstant.where(name: submitted_constants.keys)
+    constant_fields.each do |constant|
+      submitted_value = submitted_constants[constant.name]
```
Comment (omgitsbillryan) [inline app/controllers/calculator_constants_controller.rb]: There's an assumption that the params hash key represents the `CalculatorConstant.name`. This assumption isn't very readable and could start breaking down if more attributes are added to the `CalculatorConstant` model.

I think a more standard rails practice & restful design would be to lay out each attribute name explicitly:

```
"calculator_constants" => [
  {name: "foo", float_value: 999},
  {name: "bar", float_value: 888}
]
```

This also has the added benefit of allowing you to easily use rails `strong_params` if that's something you're interested in.

**Review by omgitsbillryan — APPROVED** [review]
Had a few comments you can take or leave. LGTM

File: app/controllers/calculator_constants_controller.rb [hunk 11]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+class CalculatorConstantsController < ApplicationController
+  def index
+    @calculator_constants = CalculatorConstant.all
+  end
+
+  def update
+    updated_fields = []
+    submitted_constants = params['calculator_constants'][0]
```
[the remark this answers is not in the document]
Comment (hughes-dustin) [reply app/controllers/calculator_constants_controller.rb]: I've made the change. The documentation didn't cover this scenario so I had to do some testing to confirm the functionality.

File: app/controllers/calculator_constants_controller.rb (hunk 10 above)
Comment (hughes-dustin) [reply app/controllers/calculator_constants_controller.rb]: I'm going to look into this further and try to address it in a separate PR.

**Pull request merged**